_posts/2020-11-2-kubernetes-webhooks-made-easy-with-openfaas-part-2.md

```diff
@@ -0,0 +1,291 @@
+
+```sh
+$ cat functions/fileinjector/hello-openfaas.txt
+  _  _     _ _        ___                 ___          ___
+ | || |___| | |___   / _ \ _ __  ___ _ _ | __|_ _ __ _/ __|
+ | __ / -_) | / _ \ | (_) | '_ \/ -_) ' \| _/ _` / _` \__ \
+ |_||_\___|_|_\___/  \___/| .__/\___|_||_|_|\__,_\__,_|___/
+                          |_|
+```
+
+There is one important thing we need to say about Mutating Admission webhook is that:
+
+> In a mutating admission controller webhook, mutations are performed via JSON patches. While the JSON patch standard includes a lot of intricacies that go well beyond the scope of this discussion, the Go data structure in our example as well as its usage should give the user a good initial overview of how JSON patches work:
+
+```golang
+type patchOperation struct {
+  Op    string      `json:"op"`
+  Path  string      `json:"path"`
+  Value interface{} `json:"value,omitempty"`
+}
+```
+
+For setting the field .spec.securityContext.runAsNonRoot of a pod to true, we construct the following patchOperation object:
+
+```golang
+patches = append(patches, patchOperation{
+  Op:    "add",
+  Path:  "/spec/securityContext/runAsNonRoot",
+  Value: true,
+})
+```
+
+Reminder, we will use the same workflow defined in previous post like below:
+
+* Kubernetes API -> Webhook (w/TLS) -> OpenFaaS Gateway (w/HTTP) -> OpenFaaS Function
+
+![Workflow](/images/2020-10-27-k8s-validatingwebhook-openfaas/admission-controller-phases.png)
+> Credit: [https://kubernetes.io/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/](https://kubernetes.io/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/)
+
+### Prerequisites
+##### Arkade
+
+* [arkade](https://get-arkade.dev) is The OpenFaaS community built tool for Kubernetes developers, with arkade you can easily install all necessary cli tools to your host and deploy apps to the cluster.
+
+```sh
+$ curl -sLS https://dl.get-arkade.dev | sudo sh
+```
+
+##### KinD (Kubernetes in Docker)
+
+*  Kubernetes is our recommendation for teams running at scale, but in this demo we will be using [KinD](https://kind.sigs.k8s.io/docs/user/quick-start/) for the sake of simplicity.
+
+```sh
+$ arkade get kind
+```
+
+##### kubectl
+
+* You can control your cluster using [kubectl](https://github.com/kubernetes/kubectl) CLI.
+
+```sh
+$ arkade get kubectl
+```
+
+##### faas-cli
+
+* [faas-cli](https://github.com/openfaas/faas-cli) is an official CLI for OpenFaaS , with "faas-cli" you can build and deploy functions easily.
+
+```sh
+$ arkade get faas-cli
+```
+
+### Setup
+
+### 1. Setup a Kubernetes Cluster with KinD
+
+You can start a Kubernetes cluster with KinD if you don't have one already
+
+```bash
+$ arkade get kind
+$ kind create cluster
+```
+
+### 2. Deploy OpenFaaS to our local Kubernetes Cluster with arkade:
+
+* Install a OpenFaaS
+
+```sh
+$ arkade install openfaas
+```
+
+Read the output from the installation and run the commands given to you.
+
+```sh
```

Comment:
Don't duplicate these commands, please remove. If we change them I don't want to be the one going through every post updating them :-)

Reply:
Done